### ISSUE TITLE
AndLink/OrLink/NotLink and PresentLink combinationations seem broken

### ISSUE THREAD

**Issue by einalex** [opencog/atomspace#2625]
*Case 1:* `NotLink(PresentLink(StateLink(...)))`
```
not hungry people (NotLink (PresentLink  ...) )
expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))
result: (SetLink (ConceptNode "you") (ConceptNode "hungry") (ConceptNode "me"))
```
```python
query1 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(PresentLink(StateLink(VariableNode("someone"), hungry))))


    who_isnt_hungry1 = execute_atom(atomspace, query1)
    print("\nnot hungry people (NotLink (PresentLink  ...) )")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry1)
```

*Case 2:* `OrLink(PresentLink(StateLink(...)))`
```
hungry people (OrLink (PresentLink ...))
expected: (SetLink (ConceptNode "me"))
result: (SetLink)
```
```python
query2 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                OrLink(PresentLink(StateLink(VariableNode("someone"), hungry))))

    who_is_hungry2 = execute_atom(atomspace, query2)
    print("\nhungry people (OrLink (PresentLink ...))")
    print('expected: (SetLink (ConceptNode "me"))')
    print(who_is_hungry2)
```

*Case 3:* `NotLink(AndLink(PresentLink(StateLink(...))))`
```
not hungry people (NotLink (AndLink (PresentLink ...)))
expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))
result: (SetLink (ConceptNode "you") (ConceptNode "hungry") (ConceptNode "me"))
```
```python
query3 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(AndLink(PresentLink(StateLink(VariableNode("someone"), hungry)))))

    who_isnt_hungry3 = execute_atom(atomspace, query3)
    print("\nnot hungry people (NotLink (AndLink (PresentLink ...)))")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry3)
```

*Case 4:* `NotLink(OrLink(PresentLink(StateLink(...))))`
```
not hungry people (NotLink(OrLink(PresentLink  ...)))
expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))
result: (SetLink (ConceptNode "you") (ConceptNode "hungry") (ConceptNode "me"))
```
```python
query4 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(OrLink(PresentLink(StateLink(VariableNode("someone"), hungry)))))

    who_isnt_hungry4 = execute_atom(atomspace, query4)
    print("\nnot hungry people (NotLink(OrLink(PresentLink  ...)))")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry4)
```

**Comment by einalex** [comment]
```python
#!/usr/bin/env python

from opencog.type_constructors import *
from opencog.bindlink import execute_atom, evaluate_atom
from opencog.utilities import initialize_opencog
from opencog.atomspace import AtomSpace, types

from bdi.world_state import WorldState
from bdi.knowledge_base import KnowledgeBase
from bdi.intention_recognition import IntentionRecognition



if __name__ == "__main__":
    atomspace = AtomSpace()
    initialize_opencog(atomspace)
    me = ConceptNode("me")
    you = ConceptNode("you")
    hungry = ConceptNode("hungry")

    StateLink(me, hungry)



    query = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                PresentLink(StateLink(VariableNode("someone"), hungry)))

    who_is_hungry = execute_atom(atomspace, query)
    print("\nhungry people")
    print('expected: (SetLink (ConceptNode "me"))')
    print(who_is_hungry)


    query1 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(PresentLink(StateLink(VariableNode("someone"), hungry))))


    who_isnt_hungry1 = execute_atom(atomspace, query1)
    print("\nnot hungry people (NotLink (PresentLink  ...) )")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry1)

    query2 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                OrLink(PresentLink(StateLink(VariableNode("someone"), hungry))))

    who_isnt_hungry2 = execute_atom(atomspace, query2)
    print("\nnot hungry people (OrLink (PresentLink ...))")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry2)

    query3 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(AndLink(PresentLink(StateLink(VariableNode("someone"), hungry)))))

    who_isnt_hungry3 = execute_atom(atomspace, query3)
    print("\nnot hungry people (NotLink (AndLink (PresentLink ...)))")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry3)



    query4 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                NotLink(OrLink(PresentLink(StateLink(VariableNode("someone"), hungry)))))

    who_isnt_hungry4 = execute_atom(atomspace, query4)
    print("\nnot hungry people (NotLink(OrLink(PresentLink  ...)))")
    print('expected: (SetLink (ConceptNode "you") (ConceptNode "hungry"))')
    print(who_isnt_hungry4)
```

**Comment by linas** [comment]
losing, fixed in #2626
